### PR TITLE
Bring latest Benchmarks code over from TechEmpower repo

### PR DESCRIPTION
**Pull request by dougbu**
- see individual commits for details
- mostly alignment with TechEmpower code i.e. recent fixes there

addresses:
- #251, Update benchmarks to use versions in dependencies.prop
- #263, Upgrade Npgsql from "3.2.0-*" to "3.*"

### CODE REVIEW

**Comment by cesarbs** [comment]
Can you change the title of the PR to reflect what the changes are about?

**Comment by dougbu** [comment]
@cesarbs yes, I can 😈

**Comment by dougbu** [comment]
:up::date: to match final TechEmpower code

File: build/dependencies.props [hunk 1]
```diff
@@ -14,6 +14,8 @@
     <NpgsqlVersion>3.2.0-*</NpgsqlVersion>
     <NpgsqlEntityFrameworkCorePostgreSQLVersion>1.*</NpgsqlEntityFrameworkCorePostgreSQLVersion>
 
+    <RuntimeFrameworkVersion Condition="'$(TargetFramework)'=='netcoreapp2.0'">2.0.0-*</RuntimeFrameworkVersion>
```
Comment (cesarbs) [inline build/dependencies.props]: Why do we need this?

File: src/Benchmarks/Benchmarks.csproj [hunk 2]
```diff
@@ -1,34 +1,33 @@
-﻿<Project Sdk="Microsoft.NET.Sdk.Web">
-
-  <Import Project="..\..\build\common.props" />
-
+<Project Sdk="Microsoft.NET.Sdk.Web">
   <PropertyGroup>
     <TargetFramework>netcoreapp2.0</TargetFramework>
     <OutputType>Exe</OutputType>
+    <MvcRazorCompileOnPublish>true</MvcRazorCompileOnPublish>
+    <RuntimeFrameworkVersion>2.0.0-*</RuntimeFrameworkVersion>
```
Comment (cesarbs) [inline src/Benchmarks/Benchmarks.csproj]: Don't we get this from `dependencies.props`?

File: src/Benchmarks/Benchmarks.csproj [hunk 3]
```diff
@@ -1,34 +1,33 @@
...
-    -->
-    <PackageReference Include="Npgsql" Version="$(NpgsqlVersion)" />
-    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="$(NpgsqlEntityFrameworkCorePostgreSQLVersion)" />
-    <PackageReference Include="System.Runtime.Serialization.Primitives" Version="$(CoreFxVersion)" />
+    <None Update="wwwroot/**" CopyToOutputDirectory="PreserveNewest" />
+    <None Include="appsettings.json" CopyToOutputDirectory="PreserveNewest" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Dapper" Version="1.*" />
```
Comment (cesarbs) [inline src/Benchmarks/Benchmarks.csproj]: Why are versions hardcoded again?

**Comment by davidfowl** [comment]
TechEmpower doesn't use floating versions does it?

File: build/dependencies.props (hunk 1 above)
Comment (dougbu) [inline build/dependencies.props]: To avoid relying on dependencies.props. See #251

File: src/Benchmarks/Benchmarks.csproj (hunk 2 above)
Comment (dougbu) [inline src/Benchmarks/Benchmarks.csproj]: We don't import dependencies.props. See #251

File: src/Benchmarks/Benchmarks.csproj (hunk 3 above)
Comment (dougbu) [inline src/Benchmarks/Benchmarks.csproj]: To avoid relying on dependencies.props. See #251

**Comment by dougbu** [comment]
> TechEmpower doesn't use floating versions does it?

No, it doesn't. Floating versions here and fewer supported scenarios there are the main differences between here and the TechEmpower repo.

File: build/dependencies.props (hunk 1 above)
Comment (dougbu) [inline build/dependencies.props]: Oops, wrong answer. This is here to match what we do in other repos. I bet it's there to avoid version mismatches and associated warnings.

File: src/Benchmarks/Benchmarks.csproj [hunk 4]
```diff
@@ -1,34 +1,33 @@
-﻿<Project Sdk="Microsoft.NET.Sdk.Web">
-
-  <Import Project="..\..\build\common.props" />
-
+<Project Sdk="Microsoft.NET.Sdk.Web">
   <PropertyGroup>
     <TargetFramework>netcoreapp2.0</TargetFramework>
     <OutputType>Exe</OutputType>
+    <MvcRazorCompileOnPublish>true</MvcRazorCompileOnPublish>
```
Comment (davidfowl) [inline src/Benchmarks/Benchmarks.csproj]: Why is this here

File: build/dependencies.props (hunk 1 above)
Comment (cesarbs) [inline build/dependencies.props]: Didn't realize it at first. Can you group it together with the properties at the top of the file? That's where it is in other `dependencies.props` files.

File: src/Benchmarks/Benchmarks.csproj (hunk 2 above)
Comment (cesarbs) [inline src/Benchmarks/Benchmarks.csproj]: Why don't we import it?

File: src/Benchmarks/Benchmarks.csproj (hunk 3 above)
Comment (cesarbs) [inline src/Benchmarks/Benchmarks.csproj]: Why?

File: src/Benchmarks/Benchmarks.csproj (hunk 2 above)
Comment (dougbu) [inline src/Benchmarks/Benchmarks.csproj]: @davidfowl prefers that we don't

File: src/Benchmarks/Benchmarks.csproj (hunk 4 above)
Comment (dougbu) [inline src/Benchmarks/Benchmarks.csproj]: To enable pre-compilation and to be explicit about it. Means we need fewer changes when using 1.1.x because pre-compilation wasn't the default with those versions.

File: build/dependencies.props (hunk 1 above)
Comment (dougbu) [inline build/dependencies.props]: Huh? The property settings in this file are in alphabetic order.

Comment (dougbu) [inline build/dependencies.props]: But, just realized we don't need to set `$(DapperVersion)`, `$(NpgsqlVersion)`, or `$(NpgsqlEntityFrameworkCorePostgreSQLVersion)` at all. Removing the last two will bring everything together.

File: src/Benchmarks/Benchmarks.csproj (hunk 2 above)
Comment (davidfowl) [inline src/Benchmarks/Benchmarks.csproj]: I do?

File: src/Benchmarks/Benchmarks.csproj (hunk 4 above)
Comment (davidfowl) [inline src/Benchmarks/Benchmarks.csproj]: What are we precompiling in benchmarks?

Comment (dougbu) [inline src/Benchmarks/Benchmarks.csproj]: TechEmpower scenarios include a number with MVC views.

File: src/Benchmarks/Benchmarks.csproj (hunk 2 above)
Comment (dougbu) [inline src/Benchmarks/Benchmarks.csproj]: This file previously used only properties defined in dependencies.props. So, if what you wanted was that (and I misread the issue completely), #251 was fixed before you filed it.

Comment (cesarbs) [inline src/Benchmarks/Benchmarks.csproj]: I think at the time @davidfowl filed the issue, the versions were hardcoded in the csproj. But now they are using the versions defined in `dependencies.props`, which is what he wanted. #251 had already been fixed (by me actually, but I wasn't aware there was an open issue for it).

File: build/dependencies.props (hunk 1 above)
Comment (cesarbs) [inline build/dependencies.props]: 👍

File: src/Benchmarks/Benchmarks.csproj (hunk 2 above)
Comment (davidfowl) [inline src/Benchmarks/Benchmarks.csproj]: Yep

**Comment by dougbu** [comment]
:up::date:

@davidfowl Benchmarks.csproj now uses separate properties from dependencies.props -- as we discussed.

**Comment by dougbu** [comment]
@cesarbs @davidfowl any additional comments?

File: src/Benchmarks/Middleware/DebugInfoPageMiddleware.cs [hunk 5]
```diff
@@ -42,6 +42,7 @@ public DebugInfoPageMiddleware(RequestDelegate next, IServerAddressesFeature ser
         public async Task Invoke(HttpContext httpContext)
         {
             httpContext.Response.ContentType = "text/html";
+            httpContext.Response.StatusCode = StatusCodes.Status404NotFound;
```
Comment (davidfowl) [inline src/Benchmarks/Middleware/DebugInfoPageMiddleware.cs]: Why is the status 404?

File: src/Benchmarks/Program.cs [hunk 6]
```diff
@@ -34,6 +33,7 @@ public static void Main(string[] args)
             Console.WriteLine("-----------------------");
 
             Console.WriteLine($"Current directory: {Directory.GetCurrentDirectory()}");
+            Console.WriteLine($"WebHostBuilder loading from: {typeof(WebHostBuilder).GetTypeInfo().Assembly.Location}");
```
Comment (davidfowl) [inline src/Benchmarks/Program.cs]: Why is this relevant? To show if it's using the package store or not?

File: src/Benchmarks/Program.cs [hunk 7]
```diff
@@ -59,7 +59,9 @@ public static void Main(string[] args)
                     .AddSingleton(new ConsoleArgs(args))
                     .AddSingleton<IScenariosConfiguration, ConsoleHostScenariosConfiguration>()
                     .AddSingleton<Scenarios>()
-                );
+                )
+                .UseDefaultServiceProvider(
```
Comment (davidfowl) [inline src/Benchmarks/Program.cs]: Why did we add this to the benchmarks?

File: src/Benchmarks/Startup.cs [hunk 8]
```diff
@@ -138,8 +123,6 @@ public IServiceProvider ConfigureServices(IServiceCollection services)
             {
                 services.AddResponseCaching();
             }
-
-            return services.BuildServiceProvider(validateScopes: true);
```
Comment (davidfowl) [inline src/Benchmarks/Startup.cs]: I see why

File: src/Benchmarks/Program.cs (hunk 7 above)
Comment (davidfowl) [inline src/Benchmarks/Program.cs]: Never mind.

File: build/dependencies.props [hunk 9]
```diff
@@ -2,23 +2,29 @@
...
 
     <!--
       * Use 4.4.0-* instead of $(CoreFxVersion) to address "SqlClient fails with netcoreapp2.0 on Win7/Server2008"
         * https://github.com/dotnet/corefx/issues/18406
       * Revert if and when $(CoreFxVersion) is upgraded to 4.4.0-*
     -->
     <SqlClientVersion>4.4.0-*</SqlClientVersion>
+
+    <!--
+      Following properties are specific to src/Benchmarks/Benchmarks.csproj.
```
Comment (davidfowl) [inline build/dependencies.props]: Why do we need these?

File: src/Benchmarks/Middleware/DebugInfoPageMiddleware.cs (hunk 5 above)
Comment (dougbu) [inline src/Benchmarks/Middleware/DebugInfoPageMiddleware.cs]: This middleware is only used when none of the others match i.e. nothing was found.

Besides that, I mistakenly did a manual `wrk` run with an incorrect URL. Took much longer than it should have to realize why I was seeing higher-than-MVC numbers. This prevents future mistakes like that and browsers still show all the info.

File: src/Benchmarks/Program.cs (hunk 6 above)
Comment (dougbu) [inline src/Benchmarks/Program.cs]: Yup

File: build/dependencies.props (hunk 9 above)
Comment (dougbu) [inline build/dependencies.props]: So that we can test previous releases without having to undo tracking changes in all of the web projects. This repo is far easier to automate than our fork of the TechEmpower repo.

I plan to port my branches for the existing releases from that fork here.

File: src/Benchmarks/Middleware/DebugInfoPageMiddleware.cs (hunk 5 above)
Comment (davidfowl) [inline src/Benchmarks/Middleware/DebugInfoPageMiddleware.cs]: Not that it actually mattes but it seems wrong, this is an actual valid URL.

Comment (dougbu) [inline src/Benchmarks/Middleware/DebugInfoPageMiddleware.cs]: I'd agree if this middleware only responded to specific URLs.

**Comment by dougbu** [comment]
8fa6759
